### ISSUE TITLE
autograph prevents tf.assert protection

### ISSUE THREAD

**Issue by thierryherrmann** [tensorflow/tensorflow#48699]
**System information**

    Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
    OS Platform and Distribution: Ubuntu 18.04.5 LTS
    TensorFlow installed from: binary
    TensorFlow version: 2.2.0
    Python version: 3.6.9
    CUDA/cuDNN version: 10.1.243 / 7.6.5
    GPU model and memory: NVidia Tesla V100-SXM2-32GB

**Describe the current behavior**
`tf.assert_xxx()` are not executed due to autograph (correct but unfortunate) error.

Assume we have a function taking a tensor and a bias to be added, supposedly having the same shape. We want a meaningful error message if the shapes are not the same.
First a function that doesn't do the add. Check that the error message works.

```python
@tf.function(input_signature=[tf.TensorSpec([None, None], tf.float32), tf.TensorSpec([None, None], tf.float32)])
def my_assert_func_1(x, bias):
    print('tracing...')
    tf.print('tf.shape(bias)[1]', tf.shape(bias)[1])
    tf.assert_equal(tf.shape(x)[1], tf.shape(bias)[1], message='bad shape')

    # just return x (does nothing useful)
    return x

# successful case
my_assert_func_1(tf.ones((2,5)), tf.ones((2,5)))
```
```
tracing...
tf.shape(bias)[1] 5

<tf.Tensor: shape=(2, 5), dtype=float32, numpy=
array([[1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1.]], dtype=float32)>
```
```python
# error case : the 'bad shape' message is successfully reported
my_assert_func_1(tf.ones((2,5)), tf.ones((2,7)))
```
```
tf.shape(bias)[1] 7
InvalidArgumentError:  assertion failed: [bad shape] [Condition x == y did not hold element-wise:] [x (strided_slice_1:0) = ] [5] [y (strided_slice_2:0) = ] [7]
	 [[node assert_equal_1/Assert/Assert (defined at <ipython-input-9-f5089434d4c3>:5) ]] [Op:__inference_my_assert_func_1_234]
```

Now modify the function to **actually add** the bias

```python
@tf.function(input_signature=[tf.TensorSpec([None, None], tf.float32), tf.TensorSpec([None, None], tf.float32)])
def my_assert_func_2(x, bias):
    print('tracing...')
    tf.print('tf.shape(bias)[1]', tf.shape(bias)[1])
    tf.assert_equal(tf.shape(x)[1], tf.shape(bias)[1], message='bad shape')
    # add the bias
    return x + bias

# successful case: still works
my_assert_func_2(tf.ones((2,5)), tf.ones((2,5)))
```
```
tracing...
tf.shape(bias)[1] 5

<tf.Tensor: shape=(2, 5), dtype=float32, numpy=
array([[2., 2., 2., 2., 2.],
       [2., 2., 2., 2., 2.]], dtype=float32)>
```
```python
# error case : the 'bad shape' message is NOT reported.
# Probably due to autograph that catches the error before tf.assert has a chance to execute.
# (full stacktrace in the attached gist)
my_assert_func_2(tf.ones((3,5)), tf.ones((3,7)))
```
```
tf.shape(bias)[1] 7
InvalidArgumentError:  Incompatible shapes: [3,7] vs. [3,5]
	 [[node add (defined at <ipython-input-12-b2b31a707c39>:7) ]] [Op:__inference_my_assert_func_2_284]
```

**Describe the expected behavior**
`tf_assert_xxx()` should protect against those errors and report the error message.
Not sure if this is fixable (not sure if bug or feature request, but probably hard to fix).
The consequence is that we cannot reuse in a `tf.function` libraries of custom keras layers that use `tf.assert_xxx()`.  
The attached gist shows an example of that.

**Standalone code to reproduce the issue**
https://colab.research.google.com/gist/thierryherrmann/9d253dfd32a4ab9296fadb99f5ec3651/asserts_in_tf_functions.ipynb

**Other info / logs** Include any logs or source code that would be helpful to

**Comment by bhack** [comment]
Can you try with `!pip install tf-nightly-cpu` in your colab? As It seems solved on master.

**Comment by bhack** [comment]
Ok, I got you meant having `assertion failed: [bad shape]` also in the autograph. So we still have the same behavior on master.

**Comment by bhack** [comment]
Can you try with:
```python
@tf.function(input_signature=[tf.TensorSpec([None, None], tf.float32), tf.TensorSpec([None, None], tf.float32)])
def my_assert_func_2(x, bias):
    print('tracing...')
    tf.control_dependencies((x, bias))
    tf.assert_equal(tf.shape(x)[1], tf.shape(bias)[1], message='bad shape')
    tf.print('tf.shape(bias)[1]', tf.shape(bias)[1])
    out = x + bias
    return out
```

I know that we have a note in the doc:
https://github.com/tensorflow/tensorflow/blob/90525711ff3e5e84b456c91e07323b6694647b40/tensorflow/python/framework/ops.py#L5366-L5369

But we also have:
https://github.com/tensorflow/tensorflow/blob/85c8b2a817f95a3e979ecd1ed95bff1dc1335cff/tensorflow/python/ops/check_ops.py#L646-L652

**Comment by mdanatg** [comment]
This is a known (and unfortunate) bug in the automatic control dependencies. It is almost always smart enough to wire them properly, but in this case it fails to. The main issue is that the auto control dependencies doesn't properly detect that the assert should happen before `x + bias`. Instead, it lets them run in parallel, and the add seems to win and raise an error first.

For now, a workaround is to manually add control deps:

```
with tf.control_dependencies([tf.assert_equal(tf.shape(x)[1], tf.shape(bias)[1], message='bad shape')]):
    # add the bias
    return x + bias
```

I think we should increase the priority for fixing this, because it can be unsafe. Technically,we should have a rule that a `tf.assert` should always condition the execution of any computations that depend on any tensors input to the assert itself.
cc @saxenasaurabh

**Comment by bhack** [comment]
@mdanatg I think also my workaround is ok or not?

**Comment by mdanatg** [comment]
@bhack you need to use `control_dependencies` with  a context manager -- writing it on its own has no effect. But I think we meant to represent the same idea?

**Comment by bhack** [comment]
Yes

**Comment by thierryherrmann** [comment]
```python
with tf.control_dependencies([...]):
   ...
```
is a quite a nice and non obtrusive workaround.
Thanks guys for the impressively fast response.
Up to you to close this issue or to leave it open to fix the original bug in automatic control dependencies

**Comment by bhack** [comment]
@mdanatg As I think that you have collected some well note cases do you think that we could extend the note with the currently known exceptions? 

https://github.com/tensorflow/tensorflow/blob/90525711ff3e5e84b456c91e07323b6694647b40/tensorflow/python/framework/ops.py#L5366-L5369

**Comment by mdanatg** [comment]
Yes, I believe we should. There are two broad categories that I know of:
 * uses of tf.assert (and any of its variants)
 * uses of tf.py_function which affects tf.Variables

At any rate, the note should say amount to something like "Most of the time, you shouldn't need this [...]. Only use it to manually control ordering, for example as a workaround to known issues.". Then, the docstring should include a known issues section that details these two with examples. The docstrings of tf.assert_* and tf.py_function should call that out as well.

Happy to review a PR if anyone is interested.

**Comment by bhack** [comment]
@mdanatg Yes, If you will track these features in other tickets (on github or internally?) we could reserve this to the doc change.
Can we label this with docs and contribution welcome? 

@thierryherrmann are you interested to open this small Doc PR? 
If yes check https://www.tensorflow.org/community/contribute/docs

**Comment by mdanatg** [comment]
I'm not actively working on this - if anyone from the community is interested in contributing, please add a comment here.

**Comment by Nikhil-Mudgal** [comment]
Hey @mdanatg, I am new to open source but want to start contributing to the Tensorflow repo. I am unsure how to start and work on it, Can you please provide me with some resources to start from? I will be highly obliged. Thanks in advance.

**Comment by mdanatg** [comment]
Hi @Nikhil-Mudgal , https://www.tensorflow.org/community/contribute is a good place to start.

**Comment by Nikhil-Mudgal** [comment]
Thanks @mdanatg. I checked it out but the main problem is finding a good issue for me to start on. I searched some of the issues tagged contributions welcome and good first issue; but they were either too difficult for me as a beginner in Tensorflow or I didn't understand the issue at all. Kindly help me find some issues as a beginner to try. I really want to contribute to tensorflow. 
Thanks for the help
